src/core/qgscoordinatetransform_p.cpp

```diff
@@ -242,17 +254,22 @@ QPair<projPJ, projPJ> QgsCoordinateTransformPrivate::threadLocalProjData()
     pContext = mProjContext.localData()->get();
   }
   QMap < uintptr_t, QPair< projPJ, projPJ > >::const_iterator it = mProjProjections.constFind( reinterpret_cast< uintptr_t>( pContext ) );
+#endif
 #endif
 
   if ( it != mProjProjections.constEnd() )
   {
-    QPair<projPJ, projPJ> res = it.value();
+    ProjData res = it.value();
     return res;
   }
 
   // proj projections don't exist yet, so we need to create
   locker.changeMode( QgsReadWriteLocker::Write );
 
+#if PROJ_VERSION_MAJOR>=6
+  ProjData res = proj_create_crs_to_crs( context, mSourceProjString.toUtf8(), mDestProjString.toUtf8(), nullptr );
```

Comment:
If the source CRS or target CRS is a EPSG code, then using "EPSG:xxxx" is the highly recommendd way of specifying it. PROJ strings are lossy in their ability to communicate datum information.

Reply:
Currently we use proj strings because we do something like this:

1. take the source crs as proj
2. if a user-specified datum transform needs to be applied, add the datum transform proj string (eg "+nadgrids=GDA94_GDA2020_conformal_and_distortion.gsb", or "+towgs84=-2,0,4",etc)
3. do the same with dest crs and destination datum transform
4. use these for proj_create_crs_to_crs

If instead we use the EPSG codes in proj_create_crs_to_crs, what's the process to do specify the datum transforms? Would we have to move to proj_create and use pipelines?

Reply:
And as a follow up question: how should this be done, whilst allowing for non WGS84 pivots where needed?

Reply:
> If instead we use the EPSG codes in proj_create_crs_to_crs, what's the process to do specify the datum transforms?

If you use EPSG codes in source and target, then the datum transforms will be the ones from the EPSG database that is imported in proj.db, and WGS84 will not necessarily be used as a pivot.

If you want to provide a custom datum transform, then use PROJ strings with nadgrids or towgs84.

It should be noted that the export to PROJ string of a EPSG:XXXX code no longer adds a towgs84 clause (that is if you do proj_create("EPSG:XXXX") and then proj_as_proj_string()).  You may recover some of it by using proj_crs_create_bound_crs_to_WGS84() (from proj_experimental.h) as an intermediate state, but this will still less aggressively add a towgs84 clause than what was done with the GDAL/libgeotiff CSV files. It will add one only if there's a single transformation to WGS84. For example `projinfo --boundcrs-to-wgs84 EPSG:4179` will not return a towgs84 clause, but gdalsrsinfo EPSG:4179 of GDAL < 2.5 will (it arbitrarily picked up the towgs84 clause that applied to the largest area of use)

If the user has not specified an override, then use the "EPSG:XXXX" representation.

Another thing to consider is that, even if you don't have a EPSG code, you can use WKT, and that will be much better than a PROJ string, since WKT captures the datum name, and proj_create_crs_to_crs() will be able to query the proj.db for datum transformations. So for example if you import a GDAL or OGR dataset, the OGRSpatialReference object naturally returns a WKT string. In a lot of case, the WKT has a EPSG code attached, but not necessarily.

Reply:
Hm. Ok, this is going to get very tricky. I'm going to leave it as is for now (since it's no regression from current behaviour) and will revisit at a later stage.